src/native/libs/System.Globalization.Native/pal_locale.m

```diff
@@ -734,4 +734,49 @@ int32_t GlobalizationNative_GetLocalesNative(UChar* value, int32_t length)
         return strdup([dataPath UTF8String]);
     }
 }
+
+const char* GlobalizationNative_GetDefaultLocaleNameNative(void)
+{
+    @autoreleasepool
+    {
+        NSLocale *currentLocale = [NSLocale currentLocale];
+        NSString *localeName = @"";
+
+        if (!currentLocale)
+        {
+            return strdup([localeName UTF8String]);
+        }
+
+        if ([currentLocale.languageCode length] > 0 && [currentLocale.countryCode length] > 0)
+        {
+            localeName = [NSString stringWithFormat:@"%@-%@", currentLocale.languageCode, currentLocale.countryCode];
+        }
+        else
+        {
+            localeName = currentLocale.localeIdentifier;
+        }
+
+        return strdup([localeName UTF8String]);
+    }
+}
+
+// GlobalizationNative_IsPredefinedLocaleNative returns TRUE if localeName exists in availableLocaleIdentifiers.
+// Otherwise it returns FALSE;
+
+int32_t GlobalizationNative_IsPredefinedLocaleNative(const char* localeName)
+{
+    @autoreleasepool
+    {
+        NSString *localeIdentifier = [NSString stringWithFormat:@"%s", localeName];
+        NSString *desiredLocaleIdentifier = [localeIdentifier stringByReplacingOccurrencesOfString:@"-" withString:@"_"];
```

Comment:
Why do we replace all `-` with `_`? `NSLocale availableLocaleIdentifiers` returns a `Locale ID` which [may include `-` for script designators](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15-SW9) right?

Reply:
Would that mean that `[NSLocale availableLocaleIdentifiers]` doesnt contain `Local ID`s with `_` on your system? Could it be different based on the OS version or on other devices?

Which `desiredLocaleIdentifier` contained a `-` that we needed to cast to `_` to match the `availableLocales`?

Given that the docs specify there is a possibility of including `-`, because we only cast `desiredLocaleIdentifier` to only have `_`, it seems like there is a chance to miss a match.

Reply:
Here are examples from `availableLocaleIdentifiers` in my system 
 <img width="389" alt="avialableLocales" src="https://github.com/dotnet/runtime/assets/96171496/b88ca0eb-44af-475a-a1d9-05c5e98fae0e"> 
I can see 949 Local IDs and none of them has `-`.
You are right this can be my system specific result. Will modify to not miss a match.

Reply:
Looks like that [apple doc](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15-SW9) and what I get in `availableLocaleIdentifiers` are not same. E.g.  `zh-Hant_HK` in `availableLocaleIdentifiers` is `zh_Hant_HK`, so if I change the implementation to have `localeIdentifier.ReplaceLastOccurenceOf('-', '_')` for the test case `zh-Hans-HK` it will return false but in reality it is in the `availableLocaleIdentifiers` but with the format `[language designator]_[script designator]_[region designator]` not like mentioned in the doc.

Reply:
From the code I found out that Apple calls to ICU for that information check `CFLocaleCopyAvailableLocaleIdentifiers` here https://opensource.apple.com/source/CF/CF-1153.18/CFLocale.c.auto.html. `uloc_getAvailable` calls `_canonicalize` and in that function looks like all parts in `LocalID` are appended by `'_'` https://github.com/dotnet/icu/blob/dotnet/main/icu/icu4c/source/common/uloc.cpp#L1537